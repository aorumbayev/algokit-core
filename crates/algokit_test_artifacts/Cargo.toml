[package]
name = "algokit_test_artifacts"
version = "0.1.0"
edition = "2024"
description = "Centralized contract artifacts for testing across algokit-core crates"
publish = false

[lib]
name = "algokit_test_artifacts"
path = "src/lib.rs"

[dependencies]
# No runtime dependencies - this is a test-only crate that provides static artifacts
