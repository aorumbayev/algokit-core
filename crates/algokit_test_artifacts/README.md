# AlgoKit Test Artifacts

Centralized Algorand contract artifacts for testing across algokit-core crates.

## Usage

```toml
[dev-dependencies]
algokit_test_artifacts = { path = "../algokit_test_artifacts" }
```

```rust
use algokit_test_artifacts::{constant_product_amm, sandbox};

let arc56_content = constant_product_amm::APPLICATION_ARC56;
let sandbox_contract = sandbox::APPLICATION_ARC56;
```

For full list of available contracts, inspect the contracts folder inside the crate.

## Adding New Contracts

1. Place artifact in `contracts/{contract_name}/application.arc56.json`
2. Add module and constant in `src/lib.rs`
3. Ensure contract is ARC56 compliant
