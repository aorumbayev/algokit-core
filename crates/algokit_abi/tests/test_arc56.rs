use algokit_abi::arc56_contract::Arc56Contract;
use algokit_test_artifacts::{
    arc56_struct_operations, complex_struct_test, constant_product_amm, nested_contract_calls,
    nested_struct_storage, nfd, reti, state_management_demo, template_variables, void_return_test,
    zero_coupon_bond,
};
use rstest::rstest;

#[rstest]
#[case(template_variables::APPLICATION_ARC56, "template_variables")]
#[case(state_management_demo::APPLICATION_ARC56, "state_management_demo")]
#[case(constant_product_amm::APPLICATION_ARC56, "constant_product_amm")]
#[case(nested_struct_storage::APPLICATION_ARC56, "nested_struct_storage")]
#[case(arc56_struct_operations::APPLICATION_ARC56, "arc56_struct_operations")]
#[case(complex_struct_test::APPLICATION_ARC56, "complex_struct_test")]
#[case(zero_coupon_bond::APPLICATION_ARC56, "zero_coupon_bond")]
#[case(nfd::APPLICATION_ARC56, "nfd")]
#[case(reti::APPLICATION_ARC56, "reti")]
#[case(void_return_test::APPLICATION_ARC56, "void_return_test")]
#[case(nested_contract_calls::APPLICATION_ARC56, "nested_contract_calls")]
fn test_arc56_json_roundtrip(
    #[case] artifact_content: &str,
    #[case] case_name: &str,
) -> Result<(), Box<dyn std::error::Error>> {
    // Parse JSON string to Arc56Contract
    let arc56_contract = Arc56Contract::from_json(artifact_content)?;

    // Serialize back to JSON with 4-space indentation
    let serialized_json = arc56_contract.to_json(Some(4))?;
    let parsed_json: serde_json::Value = serde_json::from_str(&serialized_json)?;

    // Snapshot test for regression detection
    insta::with_settings!({snapshot_suffix => case_name}, {
        insta::assert_json_snapshot!(parsed_json);
    });

    Ok(())
}

#[test]
fn test_arc56_state_keys_are_not_normalized() -> Result<(), Box<dyn std::error::Error>> {
    let arc56_contract = Arc56Contract::from_json(state_management_demo::APPLICATION_ARC56)?;
    let original_json: serde_json::Value =
        serde_json::from_str(state_management_demo::APPLICATION_ARC56)?;

    // Test keys that should preserve non-snake-case formatting
    let test_keys = [
        (
            &original_json["state"]["keys"]["global"],
            &arc56_contract.state.keys.global_state,
            "bytesNotInSnakeCase",
        ),
        (
            &original_json["state"]["keys"]["local"],
            &arc56_contract.state.keys.local_state,
            "localBytesNotInSnakeCase",
        ),
        (
            &original_json["state"]["keys"]["box"],
            &arc56_contract.state.keys.box_keys,
            "boxNotInSnakeCase",
        ),
    ];

    // Test box maps separately since they have a different structure (StorageMap vs StorageKey)
    let box_maps_test = (
        &original_json["state"]["maps"]["box"],
        &arc56_contract.state.maps.box_maps,
        "boxMapNotInSnakeCase",
    );

    // Verify original and parsed structures contain non-snake-case keys
    for (json_obj, parsed_map, key) in &test_keys {
        assert!(json_obj.as_object().unwrap().contains_key(*key));
        assert!(parsed_map.contains_key(*key));
    }

    // Verify box maps contain non-snake-case keys
    assert!(
        box_maps_test
            .0
            .as_object()
            .unwrap()
            .contains_key(box_maps_test.2)
    );
    assert!(box_maps_test.1.contains_key(box_maps_test.2));

    // Verify roundtrip preserves non-snake-case keys
    let exported_json = arc56_contract.to_json(Some(4))?;
    let exported_parsed: serde_json::Value = serde_json::from_str(&exported_json)?;

    for (_, _, key) in &test_keys {
        let paths = [
            &exported_parsed["state"]["keys"]["global"],
            &exported_parsed["state"]["keys"]["local"],
            &exported_parsed["state"]["keys"]["box"],
        ];
        assert!(
            paths
                .iter()
                .any(|obj| obj.as_object().unwrap().contains_key(*key))
        );
    }

    // Verify box maps preserve non-snake-case keys
    assert!(
        exported_parsed["state"]["maps"]["box"]
            .as_object()
            .unwrap()
            .contains_key(box_maps_test.2)
    );

    insta::assert_json_snapshot!(exported_parsed);
    Ok(())
}
