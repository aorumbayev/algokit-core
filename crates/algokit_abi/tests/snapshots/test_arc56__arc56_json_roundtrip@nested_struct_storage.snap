---
source: crates/algokit_abi/tests/test_arc56.rs
expression: parsed_json
---
{
  "arcs": [
    4,
    56
  ],
  "bareActions": {
    "call": [],
    "create": []
  },
  "byteCode": {
    "approval": "CiABASYCAAIAAjEYFIEGCzEZCI0MAGIAAAAAAAAAAAAAAFQAAAAAAAAAAAAAAIgAAiJDigAAiTYaAlcCADYaAReIAAIiQ4oCAIv/FigoKSgoKYv+SRUWVwYCTFCIAEZIUIgAQUhQZ4mABBUffHU2GgEXiAAEULAiQ4oBAYv/FmSJgAS4RHs2NhoAjgH/oACABOBlxHCABPRw/bY2GgCOAv+T/8QAigQDi/yL/lCM/Iv/SRWL/hcIFlcGAoz+i/1MUIz9i/yL/Yv+iQ==",
    "clear": "Cg=="
  },
  "compilerInfo": {
    "compiler": "algod",
    "compilerVersion": {
      "commitHash": "088f89dd",
      "major": 4,
      "minor": 1,
      "patch": 2
    }
  },
  "desc": "",
  "methods": [
    {
      "actions": {
        "call": [],
        "create": [
          "NoOp"
        ]
      },
      "args": [],
      "name": "createApplication",
      "returns": {
        "type": "void"
      }
    },
    {
      "actions": {
        "call": [
          "NoOp"
        ],
        "create": []
      },
      "args": [
        {
          "name": "key",
          "type": "uint64"
        },
        {
          "name": "value",
          "type": "string"
        }
      ],
      "name": "setValue",
      "returns": {
        "type": "void"
      }
    },
    {
      "actions": {
        "call": [
          "NoOp"
        ],
        "create": []
      },
      "args": [
        {
          "name": "key",
          "type": "uint64"
        }
      ],
      "name": "getValue",
      "returns": {
        "struct": "Struct2",
        "type": "((string))"
      }
    }
  ],
  "name": "NestedStructStorage",
  "source": {
    "approval": "I3ByYWdtYSB2ZXJzaW9uIDEwCmludGNibG9jayAxCmJ5dGVjYmxvY2sgMHggMHgwMDAyCgovLyBUaGlzIFRFQUwgd2FzIGdlbmVyYXRlZCBieSBURUFMU2NyaXB0IHYwLjEwNi4zCi8vIGh0dHBzOi8vZ2l0aHViLmNvbS9hbGdvcmFuZGZvdW5kYXRpb24vVEVBTFNjcmlwdAoKLy8gVGhpcyBjb250cmFjdCBpcyBjb21wbGlhbnQgd2l0aCBhbmQvb3IgaW1wbGVtZW50cyB0aGUgZm9sbG93aW5nIEFSQ3M6IFsgQVJDNCBdCgovLyBUaGUgZm9sbG93aW5nIHRlbiBsaW5lcyBvZiBURUFMIGhhbmRsZSBpbml0aWFsIHByb2dyYW0gZmxvdwovLyBUaGlzIHBhdHRlcm4gaXMgdXNlZCB0byBtYWtlIGl0IGVhc3kgZm9yIGFueW9uZSB0byBwYXJzZSB0aGUgc3RhcnQgb2YgdGhlIHByb2dyYW0gYW5kIGRldGVybWluZSBpZiBhIHNwZWNpZmljIGFjdGlvbiBpcyBhbGxvd2VkCi8vIEhlcmUsIGFjdGlvbiByZWZlcnMgdG8gdGhlIE9uQ29tcGxldGUgaW4gY29tYmluYXRpb24gd2l0aCB3aGV0aGVyIHRoZSBhcHAgaXMgYmVpbmcgY3JlYXRlZCBvciBjYWxsZWQKLy8gRXZlcnkgcG9zc2libGUgYWN0aW9uIGZvciB0aGlzIGNvbnRyYWN0IGlzIHJlcHJlc2VudGVkIGluIHRoZSBzd2l0Y2ggc3RhdGVtZW50Ci8vIElmIHRoZSBhY3Rpb24gaXMgbm90IGltcGxlbWVudGVkIGluIHRoZSBjb250cmFjdCwgaXRzIHJlc3BlY3RpdmUgYnJhbmNoIHdpbGwgYmUgIipOT1RfSU1QTEVNRU5URUQiIHdoaWNoIGp1c3QgY29udGFpbnMgImVyciIKdHhuIEFwcGxpY2F0aW9uSUQKIQpwdXNoaW50IDYKKgp0eG4gT25Db21wbGV0aW9uCisKc3dpdGNoICpjYWxsX05vT3AgKk5PVF9JTVBMRU1FTlRFRCAqTk9UX0lNUExFTUVOVEVEICpOT1RfSU1QTEVNRU5URUQgKk5PVF9JTVBMRU1FTlRFRCAqTk9UX0lNUExFTUVOVEVEICpjcmVhdGVfTm9PcCAqTk9UX0lNUExFTUVOVEVEICpOT1RfSU1QTEVNRU5URUQgKk5PVF9JTVBMRU1FTlRFRCAqTk9UX0lNUExFTUVOVEVEICpOT1RfSU1QTEVNRU5URUQKCipOT1RfSU1QTEVNRU5URUQ6CgkvLyBUaGUgcmVxdWVzdGVkIGFjdGlvbiBpcyBub3QgaW1wbGVtZW50ZWQgaW4gdGhpcyBjb250cmFjdC4gQXJlIHlvdSB1c2luZyB0aGUgY29ycmVjdCBPbkNvbXBsZXRlPyBEaWQgeW91IHNldCB5b3VyIGFwcCBJRD8KCWVycgoKLy8gY3JlYXRlQXBwbGljYXRpb24oKXZvaWQKKmFiaV9yb3V0ZV9jcmVhdGVBcHBsaWNhdGlvbjoKCS8vIGV4ZWN1dGUgY3JlYXRlQXBwbGljYXRpb24oKXZvaWQKCWNhbGxzdWIgY3JlYXRlQXBwbGljYXRpb24KCWludGMgMCAvLyAxCglyZXR1cm4KCi8vIGNyZWF0ZUFwcGxpY2F0aW9uKCk6IHZvaWQKY3JlYXRlQXBwbGljYXRpb246Cglwcm90byAwIDAKCXJldHN1YgoKLy8gc2V0VmFsdWUodWludDY0LHN0cmluZyl2b2lkCiphYmlfcm91dGVfc2V0VmFsdWU6CgkvLyB2YWx1ZTogc3RyaW5nCgl0eG5hIEFwcGxpY2F0aW9uQXJncyAyCglleHRyYWN0IDIgMAoKCS8vIGtleTogdWludDY0Cgl0eG5hIEFwcGxpY2F0aW9uQXJncyAxCglidG9pCgoJLy8gZXhlY3V0ZSBzZXRWYWx1ZSh1aW50NjQsc3RyaW5nKXZvaWQKCWNhbGxzdWIgc2V0VmFsdWUKCWludGMgMCAvLyAxCglyZXR1cm4KCi8vIHNldFZhbHVlKGtleTogdWludDY0LCB2YWx1ZTogc3RyaW5nKTogdm9pZApzZXRWYWx1ZToKCXByb3RvIDIgMAoKCS8vIG5lc3RlZF9zdHJ1Y3QuYWxnby50czoxMgoJLy8gdGhpcy5zdGF0ZShrZXkpLnZhbHVlID0geyB4OiB7IGE6IHZhbHVlIH0gfQoJZnJhbWVfZGlnIC0xIC8vIGtleTogdWludDY0CglpdG9iCglieXRlYyAwIC8vICBpbml0aWFsIGhlYWQKCWJ5dGVjIDAgLy8gIGluaXRpYWwgdGFpbAoJYnl0ZWMgMSAvLyAgaW5pdGlhbCBoZWFkIG9mZnNldAoJYnl0ZWMgMCAvLyAgaW5pdGlhbCBoZWFkCglieXRlYyAwIC8vICBpbml0aWFsIHRhaWwKCWJ5dGVjIDEgLy8gIGluaXRpYWwgaGVhZCBvZmZzZXQKCWZyYW1lX2RpZyAtMiAvLyB2YWx1ZTogc3RyaW5nCglkdXAKCWxlbgoJaXRvYgoJZXh0cmFjdCA2IDIKCXN3YXAKCWNvbmNhdAoJY2FsbHN1YiAqcHJvY2Vzc19keW5hbWljX3R1cGxlX2VsZW1lbnQKCXBvcCAvLyBwb3AgaGVhZCBvZmZzZXQKCWNvbmNhdCAvLyBjb25jYXQgaGVhZCBhbmQgdGFpbAoJY2FsbHN1YiAqcHJvY2Vzc19keW5hbWljX3R1cGxlX2VsZW1lbnQKCXBvcCAvLyBwb3AgaGVhZCBvZmZzZXQKCWNvbmNhdCAvLyBjb25jYXQgaGVhZCBhbmQgdGFpbAoJYXBwX2dsb2JhbF9wdXQKCXJldHN1YgoKLy8gZ2V0VmFsdWUodWludDY0KSgoc3RyaW5nKSkKKmFiaV9yb3V0ZV9nZXRWYWx1ZToKCS8vIFRoZSBBQkkgcmV0dXJuIHByZWZpeAoJcHVzaGJ5dGVzIDB4MTUxZjdjNzUKCgkvLyBrZXk6IHVpbnQ2NAoJdHhuYSBBcHBsaWNhdGlvbkFyZ3MgMQoJYnRvaQoKCS8vIGV4ZWN1dGUgZ2V0VmFsdWUodWludDY0KSgoc3RyaW5nKSkKCWNhbGxzdWIgZ2V0VmFsdWUKCWNvbmNhdAoJbG9nCglpbnRjIDAgLy8gMQoJcmV0dXJuCgovLyBnZXRWYWx1ZShrZXk6IHVpbnQ2NCk6IFN0cnVjdDIKZ2V0VmFsdWU6Cglwcm90byAxIDEKCgkvLyBuZXN0ZWRfc3RydWN0LmFsZ28udHM6MTYKCS8vIHJldHVybiB0aGlzLnN0YXRlKGtleSkudmFsdWUKCWZyYW1lX2RpZyAtMSAvLyBrZXk6IHVpbnQ2NAoJaXRvYgoJYXBwX2dsb2JhbF9nZXQKCXJldHN1YgoKKmNyZWF0ZV9Ob09wOgoJcHVzaGJ5dGVzIDB4Yjg0NDdiMzYgLy8gbWV0aG9kICJjcmVhdGVBcHBsaWNhdGlvbigpdm9pZCIKCXR4bmEgQXBwbGljYXRpb25BcmdzIDAKCW1hdGNoICphYmlfcm91dGVfY3JlYXRlQXBwbGljYXRpb24KCgkvLyB0aGlzIGNvbnRyYWN0IGRvZXMgbm90IGltcGxlbWVudCB0aGUgZ2l2ZW4gQUJJIG1ldGhvZCBmb3IgY3JlYXRlIE5vT3AKCWVycgoKKmNhbGxfTm9PcDoKCXB1c2hieXRlcyAweGUwNjVjNDcwIC8vIG1ldGhvZCAic2V0VmFsdWUodWludDY0LHN0cmluZyl2b2lkIgoJcHVzaGJ5dGVzIDB4ZjQ3MGZkYjYgLy8gbWV0aG9kICJnZXRWYWx1ZSh1aW50NjQpKChzdHJpbmcpKSIKCXR4bmEgQXBwbGljYXRpb25BcmdzIDAKCW1hdGNoICphYmlfcm91dGVfc2V0VmFsdWUgKmFiaV9yb3V0ZV9nZXRWYWx1ZQoKCS8vIHRoaXMgY29udHJhY3QgZG9lcyBub3QgaW1wbGVtZW50IHRoZSBnaXZlbiBBQkkgbWV0aG9kIGZvciBjYWxsIE5vT3AKCWVycgoKKnByb2Nlc3NfZHluYW1pY190dXBsZV9lbGVtZW50OgoJcHJvdG8gNCAzCglmcmFtZV9kaWcgLTQgLy8gdHVwbGUgaGVhZAoJZnJhbWVfZGlnIC0yIC8vIGhlYWQgb2Zmc2V0Cgljb25jYXQKCWZyYW1lX2J1cnkgLTQgLy8gdHVwbGUgaGVhZAoJZnJhbWVfZGlnIC0xIC8vIGVsZW1lbnQKCWR1cAoJbGVuCglmcmFtZV9kaWcgLTIgLy8gaGVhZCBvZmZzZXQKCWJ0b2kKCSsKCWl0b2IKCWV4dHJhY3QgNiAyCglmcmFtZV9idXJ5IC0yIC8vIGhlYWQgb2Zmc2V0CglmcmFtZV9kaWcgLTMgLy8gdHVwbGUgdGFpbAoJc3dhcAoJY29uY2F0CglmcmFtZV9idXJ5IC0zIC8vIHR1cGxlIHRhaWwKCWZyYW1lX2RpZyAtNCAvLyB0dXBsZSBoZWFkCglmcmFtZV9kaWcgLTMgLy8gdHVwbGUgdGFpbAoJZnJhbWVfZGlnIC0yIC8vIGhlYWQgb2Zmc2V0CglyZXRzdWI=",
    "clear": "I3ByYWdtYSB2ZXJzaW9uIDEw"
  },
  "sourceInfo": {
    "approval": {
      "pcOffsetMethod": "none",
      "sourceInfo": [
        {
          "pc": [
            0
          ],
          "source": "nested_struct.algo.ts:6",
          "teal": 1
        },
        {
          "pc": [
            1,
            2,
            3
          ],
          "source": "nested_struct.algo.ts:6",
          "teal": 2
        },
        {
          "pc": [
            4,
            5,
            6,
            7,
            8,
            9
          ],
          "source": "nested_struct.algo.ts:6",
          "teal": 3
        },
        {
          "pc": [
            10,
            11
          ],
          "source": "nested_struct.algo.ts:6",
          "teal": 15
        },
        {
          "pc": [
            12
          ],
          "source": "nested_struct.algo.ts:6",
          "teal": 16
        },
        {
          "pc": [
            13,
            14
          ],
          "source": "nested_struct.algo.ts:6",
          "teal": 17
        },
        {
          "pc": [
            15
          ],
          "source": "nested_struct.algo.ts:6",
          "teal": 18
        },
        {
          "pc": [
            16,
            17
          ],
          "source": "nested_struct.algo.ts:6",
          "teal": 19
        },
        {
          "pc": [
            18
          ],
          "source": "nested_struct.algo.ts:6",
          "teal": 20
        },
        {
          "pc": [
            19,
            20,
            21,
            22,
            23,
            24,
            25,
            26,
            27,
            28,
            29,
            30,
            31,
            32,
            33,
            34,
            35,
            36,
            37,
            38,
            39,
            40,
            41,
            42,
            43,
            44
          ],
          "source": "nested_struct.algo.ts:6",
          "teal": 21
        },
        {
          "errorMessage": "The requested action is not implemented in this contract. Are you using the correct OnComplete? Did you set your app ID?",
          "pc": [
            45
          ],
          "source": "nested_struct.algo.ts:6",
          "teal": 25
        },
        {
          "pc": [
            46,
            47,
            48
          ],
          "source": "nested_struct.algo.ts:7",
          "teal": 30
        },
        {
          "pc": [
            49
          ],
          "source": "nested_struct.algo.ts:7",
          "teal": 31
        },
        {
          "pc": [
            50
          ],
          "source": "nested_struct.algo.ts:7",
          "teal": 32
        },
        {
          "pc": [
            51,
            52,
            53
          ],
          "source": "nested_struct.algo.ts:7",
          "teal": 36
        },
        {
          "pc": [
            54
          ],
          "source": "nested_struct.algo.ts:7",
          "teal": 37
        },
        {
          "pc": [
            55,
            56,
            57
          ],
          "source": "nested_struct.algo.ts:11",
          "teal": 42
        },
        {
          "pc": [
            58,
            59,
            60
          ],
          "source": "nested_struct.algo.ts:11",
          "teal": 43
        },
        {
          "pc": [
            61,
            62,
            63
          ],
          "source": "nested_struct.algo.ts:11",
          "teal": 46
        },
        {
          "pc": [
            64
          ],
          "source": "nested_struct.algo.ts:11",
          "teal": 47
        },
        {
          "pc": [
            65,
            66,
            67
          ],
          "source": "nested_struct.algo.ts:11",
          "teal": 50
        },
        {
          "pc": [
            68
          ],
          "source": "nested_struct.algo.ts:11",
          "teal": 51
        },
        {
          "pc": [
            69
          ],
          "source": "nested_struct.algo.ts:11",
          "teal": 52
        },
        {
          "pc": [
            70,
            71,
            72
          ],
          "source": "nested_struct.algo.ts:11",
          "teal": 56
        },
        {
          "pc": [
            73,
            74
          ],
          "source": "nested_struct.algo.ts:12",
          "teal": 60
        },
        {
          "pc": [
            75
          ],
          "source": "nested_struct.algo.ts:12",
          "teal": 61
        },
        {
          "pc": [
            76
          ],
          "source": "nested_struct.algo.ts:12",
          "teal": 62
        },
        {
          "pc": [
            77
          ],
          "source": "nested_struct.algo.ts:12",
          "teal": 63
        },
        {
          "pc": [
            78
          ],
          "source": "nested_struct.algo.ts:12",
          "teal": 64
        },
        {
          "pc": [
            79
          ],
          "source": "nested_struct.algo.ts:12",
          "teal": 65
        },
        {
          "pc": [
            80
          ],
          "source": "nested_struct.algo.ts:12",
          "teal": 66
        },
        {
          "pc": [
            81
          ],
          "source": "nested_struct.algo.ts:12",
          "teal": 67
        },
        {
          "pc": [
            82,
            83
          ],
          "source": "nested_struct.algo.ts:12",
          "teal": 68
        },
        {
          "pc": [
            84
          ],
          "source": "nested_struct.algo.ts:12",
          "teal": 69
        },
        {
          "pc": [
            85
          ],
          "source": "nested_struct.algo.ts:12",
          "teal": 70
        },
        {
          "pc": [
            86
          ],
          "source": "nested_struct.algo.ts:12",
          "teal": 71
        },
        {
          "pc": [
            87,
            88,
            89
          ],
          "source": "nested_struct.algo.ts:12",
          "teal": 72
        },
        {
          "pc": [
            90
          ],
          "source": "nested_struct.algo.ts:12",
          "teal": 73
        },
        {
          "pc": [
            91
          ],
          "source": "nested_struct.algo.ts:12",
          "teal": 74
        },
        {
          "pc": [
            92,
            93,
            94
          ],
          "source": "nested_struct.algo.ts:12",
          "teal": 75
        },
        {
          "pc": [
            95
          ],
          "source": "nested_struct.algo.ts:12",
          "teal": 76
        },
        {
          "pc": [
            96
          ],
          "source": "nested_struct.algo.ts:12",
          "teal": 77
        },
        {
          "pc": [
            97,
            98,
            99
          ],
          "source": "nested_struct.algo.ts:12",
          "teal": 78
        },
        {
          "pc": [
            100
          ],
          "source": "nested_struct.algo.ts:12",
          "teal": 79
        },
        {
          "pc": [
            101
          ],
          "source": "nested_struct.algo.ts:12",
          "teal": 80
        },
        {
          "pc": [
            102
          ],
          "source": "nested_struct.algo.ts:12",
          "teal": 81
        },
        {
          "pc": [
            103
          ],
          "source": "nested_struct.algo.ts:11",
          "teal": 82
        },
        {
          "pc": [
            104,
            105,
            106,
            107,
            108,
            109
          ],
          "source": "nested_struct.algo.ts:15",
          "teal": 87
        },
        {
          "pc": [
            110,
            111,
            112
          ],
          "source": "nested_struct.algo.ts:15",
          "teal": 90
        },
        {
          "pc": [
            113
          ],
          "source": "nested_struct.algo.ts:15",
          "teal": 91
        },
        {
          "pc": [
            114,
            115,
            116
          ],
          "source": "nested_struct.algo.ts:15",
          "teal": 94
        },
        {
          "pc": [
            117
          ],
          "source": "nested_struct.algo.ts:15",
          "teal": 95
        },
        {
          "pc": [
            118
          ],
          "source": "nested_struct.algo.ts:15",
          "teal": 96
        },
        {
          "pc": [
            119
          ],
          "source": "nested_struct.algo.ts:15",
          "teal": 97
        },
        {
          "pc": [
            120
          ],
          "source": "nested_struct.algo.ts:15",
          "teal": 98
        },
        {
          "pc": [
            121,
            122,
            123
          ],
          "source": "nested_struct.algo.ts:15",
          "teal": 102
        },
        {
          "pc": [
            124,
            125
          ],
          "source": "nested_struct.algo.ts:16",
          "teal": 106
        },
        {
          "pc": [
            126
          ],
          "source": "nested_struct.algo.ts:16",
          "teal": 107
        },
        {
          "pc": [
            127
          ],
          "source": "nested_struct.algo.ts:16",
          "teal": 108
        },
        {
          "pc": [
            128
          ],
          "source": "nested_struct.algo.ts:15",
          "teal": 109
        },
        {
          "pc": [
            129,
            130,
            131,
            132,
            133,
            134
          ],
          "source": "nested_struct.algo.ts:6",
          "teal": 112
        },
        {
          "pc": [
            135,
            136,
            137
          ],
          "source": "nested_struct.algo.ts:6",
          "teal": 113
        },
        {
          "pc": [
            138,
            139,
            140,
            141
          ],
          "source": "nested_struct.algo.ts:6",
          "teal": 114
        },
        {
          "errorMessage": "this contract does not implement the given ABI method for create NoOp",
          "pc": [
            142
          ],
          "source": "nested_struct.algo.ts:6",
          "teal": 117
        },
        {
          "pc": [
            143,
            144,
            145,
            146,
            147,
            148
          ],
          "source": "nested_struct.algo.ts:6",
          "teal": 120
        },
        {
          "pc": [
            149,
            150,
            151,
            152,
            153,
            154
          ],
          "source": "nested_struct.algo.ts:6",
          "teal": 121
        },
        {
          "pc": [
            155,
            156,
            157
          ],
          "source": "nested_struct.algo.ts:6",
          "teal": 122
        },
        {
          "pc": [
            158,
            159,
            160,
            161,
            162,
            163
          ],
          "source": "nested_struct.algo.ts:6",
          "teal": 123
        },
        {
          "errorMessage": "this contract does not implement the given ABI method for call NoOp",
          "pc": [
            164
          ],
          "source": "nested_struct.algo.ts:6",
          "teal": 126
        },
        {
          "pc": [
            165,
            166,
            167
          ],
          "source": "nested_struct.algo.ts:6",
          "teal": 129
        },
        {
          "pc": [
            168,
            169
          ],
          "source": "nested_struct.algo.ts:6",
          "teal": 130
        },
        {
          "pc": [
            170,
            171
          ],
          "source": "nested_struct.algo.ts:6",
          "teal": 131
        },
        {
          "pc": [
            172
          ],
          "source": "nested_struct.algo.ts:6",
          "teal": 132
        },
        {
          "pc": [
            173,
            174
          ],
          "source": "nested_struct.algo.ts:6",
          "teal": 133
        },
        {
          "pc": [
            175,
            176
          ],
          "source": "nested_struct.algo.ts:6",
          "teal": 134
        },
        {
          "pc": [
            177
          ],
          "source": "nested_struct.algo.ts:6",
          "teal": 135
        },
        {
          "pc": [
            178
          ],
          "source": "nested_struct.algo.ts:6",
          "teal": 136
        },
        {
          "pc": [
            179,
            180
          ],
          "source": "nested_struct.algo.ts:6",
          "teal": 137
        },
        {
          "pc": [
            181
          ],
          "source": "nested_struct.algo.ts:6",
          "teal": 138
        },
        {
          "pc": [
            182
          ],
          "source": "nested_struct.algo.ts:6",
          "teal": 139
        },
        {
          "pc": [
            183
          ],
          "source": "nested_struct.algo.ts:6",
          "teal": 140
        },
        {
          "pc": [
            184,
            185,
            186
          ],
          "source": "nested_struct.algo.ts:6",
          "teal": 141
        },
        {
          "pc": [
            187,
            188
          ],
          "source": "nested_struct.algo.ts:6",
          "teal": 142
        },
        {
          "pc": [
            189,
            190
          ],
          "source": "nested_struct.algo.ts:6",
          "teal": 143
        },
        {
          "pc": [
            191
          ],
          "source": "nested_struct.algo.ts:6",
          "teal": 144
        },
        {
          "pc": [
            192
          ],
          "source": "nested_struct.algo.ts:6",
          "teal": 145
        },
        {
          "pc": [
            193,
            194
          ],
          "source": "nested_struct.algo.ts:6",
          "teal": 146
        },
        {
          "pc": [
            195,
            196
          ],
          "source": "nested_struct.algo.ts:6",
          "teal": 147
        },
        {
          "pc": [
            197,
            198
          ],
          "source": "nested_struct.algo.ts:6",
          "teal": 148
        },
        {
          "pc": [
            199,
            200
          ],
          "source": "nested_struct.algo.ts:6",
          "teal": 149
        },
        {
          "pc": [
            201
          ],
          "source": "nested_struct.algo.ts:6",
          "teal": 150
        }
      ]
    },
    "clear": {
      "pcOffsetMethod": "none",
      "sourceInfo": []
    }
  },
  "state": {
    "keys": {
      "box": {},
      "global": {},
      "local": {}
    },
    "maps": {
      "box": {},
      "global": {
        "state": {
          "keyType": "uint64",
          "valueType": "Struct2"
        }
      },
      "local": {}
    },
    "schema": {
      "global": {
        "bytes": 10,
        "ints": 0
      },
      "local": {
        "bytes": 0,
        "ints": 0
      }
    }
  },
  "structs": {
    "Struct2": [
      {
        "name": "x",
        "type": [
          {
            "name": "a",
            "type": "string"
          }
        ]
      }
    ]
  }
}
